### PR TITLE
perf(messages): write-behind buffer for messageSecret persistence

### PR DESCRIPTION
**Pull request by jlucaso1**
## Problem

Every inbound message carrying a `messageSecret` (polls, events, bot prompts, and with the default `Managed` policy any secret-bearing text) paid a synchronous SQLite upsert inside the per-chat receive lane: `dispatch_parsed_message` awaited `maybe_capture_inbound_msg_secret` -> `put_msg_secrets` -> an immediate transaction, before the ack that drains the server's offline queue and before the consumer received `Event::Message`. Under an offline drain or a group fan-in burst that is one transaction per message on the critical path.

The ordering is load-bearing, though: an add-on (edit, reaction, poll vote) arriving right after its parent must find the parent's secret, so the write could not simply become fire-and-forget.

## Change

A `MsgSecretWriteBuffer` splits visibility from durability:

- A capture inserts into an in-memory pending map synchronously (no await), so the per-chat lane ordering guarantee is preserved by construction, and then schedules a detached drain task.
- The drain batches everything pending into one `put_msg_secrets` call, looping until the buffer is empty. A burst of captures coalesces into a single transaction.
- Every secret read (`maybe_decrypt_secret_encrypted_message` primary and alternate lookups, the msmsg bot-reply path, and the LID/PN alternate helper) checks the buffer before the store, so a not-yet-flushed entry is never invisible. Entries leave the buffer only after the backend write returns, so a reader always finds a secret either pending or persisted, never neither.
- A failed write warns and drops, the same data-loss semantics the previously awaited error path had. Secrets for a given key never change, so the post-write removal cannot discard a newer value.
- The outbound send-time secret store (`persist_outbound_msg_secret`) goes through the same buffer, and msmsg replies read buffer-first, covering the gap.

The durability tradeoff is explicit: the window between ack and flush can lose secrets on a process crash (recoverable via the app resolver). WA Web has the same window, since its IndexedDB persistence is asynchronous.

## Benchmark

The receive-lane win is structural: zero awaited I/O in the capture path (the function is no longer async at the write site). The batching is demonstrated empirically: `burst_coalesces_into_one_batch` queues 20 captures and asserts exactly 1 `put_msg_secrets` batch reached the store, where the old path issued 20 immediate transactions.

## Tests

- `read_your_write_before_flush`: a queued entry is readable before the drain task ever polls (deterministic on the current-thread test runtime) and lands in the backend after.
- `burst_coalesces_into_one_batch`: 20 captures -> 1 batch, all rows present.
- `entries_queued_during_drain_are_flushed`: entries racing an in-flight drain are picked up, never lost.
- `addon_decrypts_right_after_capture_without_flush`: the receive-lane semantic guard, capture immediately followed by a secret-encrypted edit referencing it, no flush in between.
- Existing capture/persist tests now wait deterministically for the flush before asserting on the store (`wait_flushed`, test-only).
- Full workspace suite green, `cargo clippy --all-targets -- -D warnings` clean.

## Breaking

None. Note: this PR and #830 both touch `src/message/msg_secret.rs`; whichever merges second needs a trivial rebase.

### CODE REVIEW

File: src/msg_secret_buffer.rs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+                // Same semantics as the previously awaited write: warn + drop.
+                log::warn!("failed to persist messageSecrets: {e:?}");
+            }
+            self.flushed_batches.fetch_add(1, Ordering::Relaxed);
+            {
+                // A secret for a given (chat, sender, id) never changes, so
+                // removing by key cannot discard a newer value.
+                let mut pending = self.pending.lock().unwrap_or_else(|p| p.into_inner());
+                for key in keys {
+                    pending.remove(&key);
```
[the remark this answers is not in the document]
Comment (jlucaso1) [reply src/msg_secret_buffer.rs]: You're right, the edit recapture path stores a new secret under the same parent id, so the immutability assumption didn't hold. Fixed in ca321be: the post-flush cleanup now removes a key only when the pending value still matches what was written, so a racing refresh stays for the next drain iteration. Regression test added (refresh_queued_during_flush_survives_removal).

File: src/msg_secret_buffer.rs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    fn schedule_drain(self: &Arc<Self>) {
+        if self.drain_in_flight.swap(true, Ordering::AcqRel) {
+            return;
+        }
+        let buffer = Arc::clone(self);
+        self.runtime
+            .spawn(Box::pin(async move {
+                buffer.drain_loop().await;
+            }))
+            .detach();
```
[the remark this answers is not in the document]
Comment (jlucaso1) [reply src/msg_secret_buffer.rs:129]: Agreed. Fixed in 861d0ca: the buffer now has a production flush() and disconnect() drains it right before the device-state flush, so a clean exit after a capture no longer loses the only copy. Covered by explicit_flush_drains_everything.

[the remark this answers is not in the document]
Comment (jlucaso1) [reply src/msg_secret_buffer.rs:129]: Done in 861d0ca: flush() is a real production API now (idempotent against a concurrent drain) and the disconnect path awaits it before returning, next to persistence_manager.flush().

File: src/client/lifecycle.rs [hunk 3]
```diff
@@ -524,6 +528,9 @@ impl Client {
             .await;
         self.notify_connection_shutdown();
 
+        // The write-behind secret drain is detached; a clean exit right after
+        // a capture must not lose the only copy.
+        self.msg_secret_buffer.flush().await;
```
[the remark this answers is not in the document]
Comment (jlucaso1) [reply src/client/lifecycle.rs]: Fair point. In 6aa9b8c the disconnect flush moved to after the connection teardown (transport closed, connection state cleaned), so receive work that was in flight when disconnect started has had its captures queued by then, and flush() loops until nothing is pending. A producer that somehow queues after that still lands on the detached drain; that residual window is process-exit-immediately-after-disconnect with a lane task mid-processing, which is the same class of in-flight async work the rest of shutdown already tolerates.

File: src/msg_secret_buffer.rs [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        self.flushed_batches.fetch_add(1, Ordering::Relaxed);
+        self.finish_batch(&batch);
+        true
+    }
+
+    /// Drain everything pending before returning. For graceful shutdown: the
+    /// detached drain task is not awaited anywhere, so disconnect calls this
+    /// to make sure a just-captured secret is not lost on a clean exit.
+    pub(crate) async fn flush(&self) {
+        while self.flush_pending_once().await {}
```
[the remark this answers is not in the document]
Comment (jlucaso1) [reply src/msg_secret_buffer.rs:177]: You're right, two competing upserts had no ordering guarantee. Fixed in 6aa9b8c: every backend write (drain and flush) snapshots the pending map under a shared async lock, so a later writer always carries values at least as new as any earlier in-flight write and a stale snapshot can never land last.

File: src/msg_secret_buffer.rs [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+        if entries.is_empty() {
+            return;
+        }
+        {
+            let mut pending = self.pending.lock().unwrap_or_else(|p| p.into_inner());
+            for entry in entries {
+                let key = (
+                    entry.chat.clone(),
+                    entry.sender.clone(),
+                    entry.msg_id.clone(),
```
[the remark this answers is not in the document]
Comment (jlucaso1) [reply src/msg_secret_buffer.rs:81]: Good catch. queue() now applies the same canonical merge the backend's on-conflict clause uses (merge_msg_secret_expiry and merge_msg_secret_message_ts from wacore) before coalescing a duplicate, so never-expire wins, windows never shrink and a known parent time survives a later unknown one. Fixed with a regression test in the latest push.

File: src/msg_secret_buffer.rs [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        self.flushed_batches.fetch_add(1, Ordering::Relaxed);
+        self.finish_batch(&batch);
+        true
+    }
+
+    /// Drain everything pending before returning. For graceful shutdown: the
+    /// detached drain task is not awaited anywhere, so disconnect calls this
+    /// to make sure a just-captured secret is not lost on a clean exit.
+    pub(crate) async fn flush(&self) {
+        while self.flush_pending_once().await {}
```
[the remark this answers is not in the document]
Comment (jlucaso1) [reply src/msg_secret_buffer.rs:177]: You're right, and the unrecoverable case is real: a backlog message acked after the shutdown flush is never redelivered, so its secret had no durable copy. Fixed in the latest push by sealing the buffer at terminal disconnect, right before the final flush: a sealed queue() writes inline before returning (the old synchronous semantics), and the pending mutex orders the seal against the flush snapshot so an insert the snapshot missed always observes sealed and writes inline. The normal hot path only pays an atomic flag load. Covered by sealed_queue_writes_inline.
